### PR TITLE
Add helper to assert writeability to x-content

### PR DESCRIPTION
**Pull request by rjernst**
When writing generic objects to x-content the value may cause an error if XContentBuilder does not know how to understand the concrete object type. This commit adds a new helper method, similar to
StreamOutput.checkWriteable, which validates the type of an object (and any inner objects if it is a collection) are writeable to x-content.

### CODE REVIEW

File: libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -1302,4 +1304,41 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
...
+            return;
+        }
+
+        if (value instanceof Object[] array) {
+            for (Object v : array) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof Map<?,?> map) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                ensureToXContentable(entry.getKey());
```
Comment (tvernum) [inline libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java:1323]: Doesn't `unknownValue` require that the key is a `String`, and not just any writable value?

File: libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java [hunk 2]
```diff
@@ -1302,4 +1304,41 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
...
+            // Path implements Iterable<Path> and causes endless recursion and a StackOverFlow if treated as an Iterable here
+            return;
+        } else if (value instanceof Iterable<?> iterable) {
+            // Iterable also implicitly handles Set and List
+            for (Object v : iterable) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof ToXContent) {
+            return;
+        } else {
```
Comment (tvernum) [inline libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java:1336]: Is it intentional that there is no `Enum` case? (`unknownValue` has one)

Comment (rjernst) [reply libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java:1336]: Not intentional. Added.

File: libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java (hunk 1 above)
Comment (rjernst) [reply libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java]: Yep, added a check for String and a test.

File: libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java [hunk 3]
```diff
@@ -1302,4 +1303,45 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
...
+
+        if (value instanceof Object[] array) {
+            for (Object v : array) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof Map<?, ?> map) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                if (entry.getKey() instanceof String == false) {
+                    throw new IllegalArgumentException(
+                        "Cannot write non-String map key [" + entry.getKey().getClass().getCanonicalName() + "] to x-content"
```
Comment (tvernum) [inline libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java:1325]: It's possible for a class to have a `null` canonical name (e.g. anonymous classes).
Should this just be `getName()` ?

**Review by tvernum — APPROVED** [review]
LGTM.
I'll let you decide whether to stick with `getCanonicalName()`

**Pull request merged**